### PR TITLE
feat: Add new fields in CheCluster CR to configure Git Services

### PR DESCRIPTION
**Pull request by tolusha**
Signed-off-by: Anatolii Bazko <abazko@redhat.com>

### What does this PR do?
* Add new fields in CheCluster CR to configure Git Services.
* Automatically sets the required labels and annotations to Git integration secrets

### Screenshot/screencast of this PR

screenshot-console-openshift-console apps ci-ln-zr383rt-72292 origin-ci-int-gce dev rhcloud com-2022 09 07-13_46_10
screenshot-console-openshift-console apps ci-ln-zr383rt-72292 origin-ci-int-gce dev rhcloud com-2022 09 07-13_45_59
screenshot-console-openshift-console apps ci-ln-zr383rt-72292 origin-ci-int-gce dev rhcloud com-2022 09 07-13_45_43

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/21577

### How to test this PR?

Catalog source image to test: `docker.io/abazko/catalog:21577`

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [x] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by tolusha** [comment]
@l0rd 
Could you have a look pls?

File: api/v2/checluster_types.go [hunk 1]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	GitLab []GitLabService `json:"gitlab,omitempty"`
+	// +optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="BitBucket"
+	BitBucket []BitBucketService `json:"bitbucket,omitempty"`
+}
+
+// GitHubService enables users to work with a remote Git repository that is hosted on GitHub.
+type GitHubService struct {
+	// Kubernetes secret, that contains Base64-encoded GitHub OAuth Client id and GitHub OAuth Client secret data.
+	// The GitHub OAuth data must be stored in the Kubernetes secret in `id` and `secret` keys respectively.
```
Comment (l0rd) [inline api/v2/checluster_types.go:564]: I think we should provide a link to the documentation. Not sure if markdown is rendered correctly though...
```suggestion
	// The GitHub OAuth data must be stored in the Kubernetes secret in `id` and `secret` keys respectively.
	// See [Eclipse Che Administration Guide](https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-github/) for details.
```

File: api/v2/checluster_types.go [hunk 2]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
 
+type CheClusterGitServices struct {
+	// +optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="GitHub"
+	GitHub []GitHubService `json:"github,omitempty"`
+	// +optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="GitLab"
+	GitLab []GitLabService `json:"gitlab,omitempty"`
+	// +optional
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="BitBucket"
```
Comment (l0rd) [inline api/v2/checluster_types.go:557]: Bitbucket is not camel case, it's one word (https://bitbucket.org/). It's up to you to update the `struct` field or not but I think that updating the `displayName` is important.
```suggestion
	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Bitbucket"
```

File: api/v2/checluster_types.go [hunk 3]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	BitBucket []BitBucketService `json:"bitbucket,omitempty"`
+}
+
+// GitHubService enables users to work with a remote Git repository that is hosted on GitHub.
```
Comment (l0rd) [inline api/v2/checluster_types.go:561]:
```suggestion
// GitHubService enables users to work with repositories hosted on GitHub (GitHub.com or GitHub Enterprise).
```

File: api/v2/checluster_types.go [hunk 4]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+type GitHubService struct {
+	// Kubernetes secret, that contains Base64-encoded GitHub OAuth Client id and GitHub OAuth Client secret data.
+	// The GitHub OAuth data must be stored in the Kubernetes secret in `id` and `secret` keys respectively.
+	// +kubebuilder:validation:Required
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:io.kubernetes:Secret"
+	SecretName string `json:"secretName"`
+}
+
+type GitLabService struct {
```
Comment (l0rd) [inline api/v2/checluster_types.go:570]:
```suggestion
// GitLabService enables users to work with repositories hosted on GitLab (gitlab.com or self-hosted).
type GitLabService struct {
```

File: api/v2/checluster_types.go [hunk 5]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	// +kubebuilder:validation:Required
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:io.kubernetes:Secret"
+	SecretName string `json:"secretName"`
+	// GitLab server endpoint URL.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:="https://gitlab.com"
+	Endpoint string `json:"endpoint"`
+}
+
+type BitBucketService struct {
```
Comment (l0rd) [inline api/v2/checluster_types.go:582]:
```suggestion
// BitBucketService enables users to work with repositories hosted on Bitbucket (bitbucket.org or self-hosted).
type BitBucketService struct {
```

File: api/v2/checluster_types.go [hunk 6]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+type GitHubService struct {
+	// Kubernetes secret, that contains Base64-encoded GitHub OAuth Client id and GitHub OAuth Client secret data.
+	// The GitHub OAuth data must be stored in the Kubernetes secret in `id` and `secret` keys respectively.
+	// +kubebuilder:validation:Required
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:io.kubernetes:Secret"
+	SecretName string `json:"secretName"`
+}
+
+type GitLabService struct {
+	// Kubernetes secret, that contains Base64-encoded GitHub Application id and GitHub Application Client secret data.
```
Comment (l0rd) [inline api/v2/checluster_types.go:571]:
```suggestion
	// Kubernetes secret, that contains Base64-encoded GitLab Application id and GitLab Application Client secret data.
```

File: api/v2/checluster_types.go [hunk 7]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	// The GitHub Application data must be stored in `id` and `secret` keys of the Kubernetes secret respectively.
```
Comment (l0rd) [inline api/v2/checluster_types.go:572]:
```suggestion
	// The Gitlab Application data must be stored in `id` and `secret` keys of the Kubernetes secret respectively.
	// See [Eclipse Che Administration Guide](https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/) for details.
```

File: api/v2/checluster_types.go [hunk 8]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:io.kubernetes:Secret"
+	SecretName string `json:"secretName"`
+	// GitLab server endpoint URL.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:="https://gitlab.com"
+	Endpoint string `json:"endpoint"`
+}
+
+type BitBucketService struct {
+	// Kubernetes secret, that contains Base64-encoded BitBucket OAuth 1.0 or OAuth 2.0 data.
```
Comment (l0rd) [inline api/v2/checluster_types.go:583]:
```suggestion
	// Kubernetes secret, that contains Base64-encoded Bitbucket OAuth 1.0 or OAuth 2.0 data.
```

File: api/v2/checluster_types.go [hunk 9]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	// For OAuth 1.0: private key, BitBucket Application link consumer key and BitBucket Application link shared secret must be stored
```
Comment (l0rd) [inline api/v2/checluster_types.go:584]:
```suggestion
	// For OAuth 1.0: private key, Bitbucket Application link consumer key and Bitbucket Application link shared secret must be stored
```

File: api/v2/checluster_types.go [hunk 10]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	// in `private.key`, `consumer.key` and `shared_secret` keys of the Kubernetes secret respectively.
+	// For OAuth 2.0: BitBucket OAuth consumer key and BitBucket OAuth consumer secret must be stored
```
Comment (l0rd) [inline api/v2/checluster_types.go:586]:
```suggestion
	// For OAuth 2.0: BitBucket OAuth consumer key and Bitbucket OAuth consumer secret must be stored
```

File: api/v2/checluster_types.go [hunk 11]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	// in `id` and `secret` keys of the Kubernetes secret respectively.
```
Comment (l0rd) [inline api/v2/checluster_types.go:587]:
```suggestion
	// in `id` and `secret` keys of the Kubernetes secret respectively.
	// See [Eclipse Che Administration Guide](https://www.eclipse.org/che/docs/stable/administration-guide/configuring-oauth-1-for-bitbucket/) for details.
```
@tolusha @vinokurig I don't see OAuth 2 bitbucket in Che documentation.

**Review by l0rd — COMMENTED** [review]
@tolusha good work. I have added a few suggestions.

File: api/v2/checluster_types.go [hunk 12]
```diff
@@ -541,6 +546,54 @@ type PodSecurityContext struct {
...
+	BitBucket []BitBucketService `json:"bitbucket,omitempty"`
+}
+
+// GitHubService enables users to work with a remote Git repository that is hosted on GitHub.
+type GitHubService struct {
+	// Kubernetes secret, that contains Base64-encoded GitHub OAuth Client id and GitHub OAuth Client secret data.
+	// The GitHub OAuth data must be stored in the Kubernetes secret in `id` and `secret` keys respectively.
+	// +kubebuilder:validation:Required
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors="urn:alm:descriptor:io.kubernetes:Secret"
+	SecretName string `json:"secretName"`
```
Comment (l0rd) [inline api/v2/checluster_types.go:567]: @vinokurig why we don't we need the endpoint for GitHub? How does it work? Do we assume that's a github server if it's not among gitlab/bitbucket endpoints?

File: api/v2/checluster_types.go (hunk 11 above)
Comment (vinokurig) [reply api/v2/checluster_types.go:594]: There is a PR for it https://github.com/eclipse-che/che-docs/pull/2410

File: api/v2/checluster_types.go (hunk 12 above)
Comment (vinokurig) [reply api/v2/checluster_types.go:571]: As for the GitHub case, we don't support the self hosted GitHub server endpoint. It means that the SAAS version and the cloud version of GitHub enterprise always use the `github.com` endpoint. So we don't need to configure it yet.

**Comment by tolusha** [comment]
This PR has to be updated to mention Bitbucket SAAS OAuth flow 
Depends on https://github.com/eclipse-che/che-docs/pull/2410

**Pull request merged**